contrib/qubes/src/vms/fwupd_download_updates.py

```diff
@@ -0,0 +1,178 @@
+METADATA_URL_JCAT = "https://fwupd.org/downloads/firmware.xml.gz.jcat"
+
+
+class DownloadData(FwupdVmCommon):
+    def _decrypt_update_url(self, url):
+        self.dec_url = url
+        if "--and--" in url:
+            self.dec_url = self.dec_url.replace("--and--", "&")
+            self.arch_name = "untrusted.cab"
+        if "--or--" in url:
+            self.dec_url = self.dec_url.replace("--or--", "|")
+            self.arch_name = "untrusted.cab"
+        if "--hash--" in url:
+            self.dec_url = self.dec_url.replace("--hash--", "#")
+            self.arch_name = "untrusted.cab"
+        if "%20" in url:
+            self.arch_name = "untrusted.cab"
+
+    def _download_metadata_file(self):
+        """Download metadata file"""
+        if self.custom_url is None:
+            metadata_url = METADATA_URL
+        else:
+            metadata_url = self.custom_url
+        cmd_metadata = [
+            "wget",
+            "-P",
+            FWUPD_VM_METADATA_DIR,
+            metadata_url
+        ]
+        p = subprocess.Popen(cmd_metadata)
+        p.wait()
+        if p.returncode != 0:
+            raise Exception("fwudp-qubes: Downloading metadata file failed")
+        if not os.path.exists(self.metadata_file):
+            raise FileNotFoundError(
+                "fwudp-qubes: Downloaded metadata file does not exist"
+            )
+
+    def _download_metadata_jcat(self):
+        """Download metadata jcat signature"""
+        if self.custom_url is None:
+            metadata_url = METADATA_URL
+        else:
+            metadata_url = self.custom_url
+        cmd_metadata = [
+            "wget",
+            "-P",
+            FWUPD_VM_METADATA_DIR,
+            f"{metadata_url}.jcat"
+        ]
+        p = subprocess.Popen(cmd_metadata)
+        p.wait()
+        if p.returncode != 0:
+            raise Exception("fwudp-qubes: Downloading metadata file failed")
+        if not os.path.exists(f"{self.metadata_file}.jcat"):
+            raise FileNotFoundError(
+                "fwudp-qubes: Downloaded metadata file does not exist"
+            )
+
+    def download_metadata(self, url=None):
+        """Downloads default metadata and its signatures"""
+        if url is not None:
+            self.custom_url = url
+            custom_metadata_name = url.replace(
+                FWUPD_DOWNLOAD_PREFIX,
+                ""
+            )
+            self.metadata_file = os.path.join(
+                FWUPD_VM_METADATA_DIR,
+                custom_metadata_name
+            )
+        else:
+            self.custom_url = None
+            self.metadata_file = os.path.join(
+                FWUPD_VM_METADATA_DIR,
+                "firmware.xml.gz"
+            )
+        self.validate_dirs()
+        self._download_metadata_file()
+        self._download_metadata_jcat()
+
+    def _check_shasum(self, file_path, sha):
+        """Compares computed SHA1 checksum with `sha` parameter.
```

Comment:
I think I have to waive all SHA1 use in RHEL; could we use SHA256 instead?

Reply:
I replaced SHA1, with SHA256. I can see that some firmware updates don't have SHA256 checksum - e.g. ColorHug2 which I use to test the wrapper. Is it possible to add the SHA256 to this device? Should I add sanity checks if there is SHA256 in the update metadata?